### DISCUSSION TITLE
Using React Flow for Backend Workflow Automation

### DISCUSSION THREAD

**Discussion by jamalsoueidan** [xyflow/xyflow#4616] [Q&A] [answered]
I'm working on a project where I use React Flow to design interactive message flows for chatbot. The frontend allows users to create and visualize these flows using nodes and edges, and I’m considering leveraging React Flow’s logic directly in the backend to manage these workflows without rendering the UI.

**Current Setup:**

I have React Flow configured on the frontend for users to build and visualize workflows.
The flows are stored in db of nodes and edges.
Each flow represents a sequence of actions, like sending an interactive list, or sending a simple message.

**When a user interacts with the chatbot, I need the backend to:**

Process user input by moving through the flow's nodes and edges, executing the appropriate actions (e.g., sending messages, calling APIs).
Manage the user's state within the flow (e.g., knowing which node they are on and what the next node).
I’m considering using React Flow’s data structures and logic to manage this in the backend:

Is it possible to use React Flow in a non-rendering context, purely for backend logic? For instance, using React Flow’s hooks and utilities to navigate nodes and edges.

**Comment by ckpearson** [comment, marked as answer]
So we actually do something similar on the backend, except we don't use react flow for it, and you probably don't need to either.

You just want to take your nodes and edges and build a structure representing the graph, and provided you don't allow cycles, you could perform a topological sort so you get the order to process your nodes in.

We keep a big bag of the outputs from each node, and then as we hit the next node to process because we ordered them properly first, we know its inputs are available for consumption.

**Reply by jamalsoueidan** [reply]
My situation is little bit different since the flow is executed one node at the time, and not the whole flow at once. Each user interaction move the user from one node to next.

**Reply by ckpearson** [reply]
Oh yeah ours too. Basically once the topological sort is done, it's then just a big foreach loop.

If your flow isn't even a graph, but just a linear sequence of steps that's even easier, just order them based on their input steps and walk through them.

**Reply by jamalsoueidan** [reply]
Great tips!

My flow is something like this:

start (condition if it should start this flow for the user)
A (send a list of choices, user requires to pick, before continue)
B | C (depending on previous choice, pick the next node, and execute)

**Reply by ckpearson** [reply]
Our system is pure data flow, so no conditional branching which makes things simpler, but is sort of achieved in a pseudo-way by allowing steps to either not output data, or we have a typed data layer on top where steps can pipe out "truthy" values (along with other things).

So you likely want to come up with some way of representing the connections between your nodes, and then the data specific to them so something like:

```json
[
  {
     "id": "choice-node-1",
     "type": "choice",
     "choices": [ 
         { "id": "1", "value": "Choice 1" , "nextNode": "b"},
         { "id": "2", "value": "Choice 2" , "nextNode": "c"} 
      ]
  },
  {
    "id": "node-b",
    "type": "register.basic"
  },
  {
    "id": "node-c",
    "type": "register.premium"
  }
]
```

You'd want to adapt this to your own mechanism, but basically:

1. Your trigger knows what its conditions are, and what the "start" node should be (e.g. `choice-node-1`)
2. Choice node 1 would present the choices to the user and it knows which node each choice leads to
3. Your final terminating nodes would either lead to a "stop" node, or you could just assume if there's no `nextNode` then they are the end of the process.

Your processing logic then would be something like:

```
var myState = GetCurrentWorkflowState();
var nextNode = myState.GetNextNodeToExecute();
nextNode.Process(myState);
UpdateWorkflowState(myState);
```

This way you could process the flow node-by-node, and on every state update decide what to do next, either:

1. Trigger the process to use the next node
2. You're done!

**Reply by jamalsoueidan** [reply]
> Our system is pure data flow, so no conditional branching which makes things simpler, but is sort of achieved in a pseudo-way by allowing steps to either not output data, or we have a typed data layer on top where steps can pipe out "truthy" values (along with other things).
> 
> So you likely want to come up with some way of representing the connections between your nodes, and then the data specific to them so something like.....

The UI frontend is built, and I did exactly like you mentiond.
![chrome_WuyHKGOgk6](https://github.com/user-attachments/assets/872936e0-49be-4886-bf13-4c1e3492976f)

The process is somehow done now, but every node PUSH out a message, and EXPECT a message in, before proceding to next node. 

https://github.com/user-attachments/assets/f3dd790c-ec9c-42c1-b924-9e89c4664847

If I keep using the reactflow data in my backend data, I can make the client see the process working live in the reactflow ui, which I think would be cool.

Is your project shared anywhere?

**Reply by ckpearson** [reply]
My project isn't live anywhere sadly, I'd recommend against running this within the flow editor as this ultimately is going to need to be something running in the background persistently, and you probably don't want a background task relying on UI code.

If you wanted to be able to show the user what the flow is getting up to in-editor, my recommendation would be to have your flow processor code store information that indicates what it's up to, what each step's value was etc, and then periodically communicate this state to the frontend - could be a polling request from the frontend, or some sort of push like SignalR / server-sent-events etc etc.

Then in your frontend editor, you could receive this execution state data, and use it to update your nodes with information to render this.

**Reply by jamalsoueidan** [reply]
Lets go....the next would be to allow the user to test the flow from specific node instead of from the start all the time...

In the future I can show all the events in realtime, somekind of avatar that jump from node to node, to visual where all the users is moving, just for fun...

https://github.com/user-attachments/assets/f4870bde-13f1-48e7-9c86-c40abc1f51ab

**Reply by ckpearson** [reply]
Bloody hell that was quick, but hell yeah! If you wanted to get super fancy and really help power realtime display / timeline replay, you could switch to treating the play state of a flow as a timeline of events:

```json
[
  { type: "message_received", node_id: "start", payload: { message: "hey" }, timestamp: XXXXXXXX },
  { type: "choice_made", node_id: "buttons", payload: { choice: "B" }, timestamp: XXXXX}
]
```

And so on and so forth, then all you would have to do is periodically send fresher events along to the client, or have it poll and ask for events fresher than a given timestamp / sequenced ID.

Super super fancy would be to utilise something like a CRDT to have the state changes communicated between client / server, or even multiple users in a collaborative editor.

Glad to see you got this working!

**Reply by jamalsoueidan** [reply]
> Bloody hell that was quick, but hell yeah! If you wanted to get super fancy and really help power realtime display / timeline replay, you could switch to treating the play state of a flow as a timeline of events:
> 
> ```json
> [
>   { type: "message_received", node_id: "start", payload: { message: "hey" }, timestamp: XXXXXXXX },
>   { type: "choice_made", node_id: "buttons", payload: { choice: "B" }, timestamp: XXXXX}
> ]
> ```
> 
> And so on and so forth, then all you would have to do is periodically send fresher events along to the client, or have it poll and ask for events fresher than a given timestamp / sequenced ID.
> 
> Super super fancy would be to utilise something like a CRDT to have the state changes communicated between client / server, or even multiple users in a collaborative editor.
> 
> Glad to see you got this working!

Thanks, thats good idea, but lets wait, I gotta finish the nodes to work with the backend, this is for whatsapp cloud api, so you can setup a flow for your business whenever someone contact you, you can let them go through steps, maybe find lost items, or get the location of the store etc.
